python/lsst/resources/_resourcePath.py

```diff
@@ -206,7 +211,7 @@ def __new__(
             # We invoke __new__ again with str(self) to add a scheme for
             # forceAbsolute, but for the others that seems more likely to paper
             # over logic errors than do something useful, so we just raise.
-            if forceDirectory and not uri.dirLike:
+            if forceDirectory and not uri.isdir():
```

Comment:
If `forceDirectory and uri.dirLike is False` should definitely raise an exception.

The question is what you do if the URI you are given doesn't actually know it's a directory or not. If you are doing

```
resource_copy = ResourcePath(resource, forceDirectory=True)
```
and `resource` doesn't know if it's a directory or not, then things can get dicey if it is in fact a local file. But yes, if this happens to be a relative path that will soon be attached to an S3 root URI then checking the file system for directory-ness is not great. Maybe we only do this check if it's a `file` URI since then we really can tell and now the `os.stat` call will only happen if it hasn't already been cached or pre-declared. Then for schemeless we cross fingers and hope you know what you are doing.

We do have a problem in that we can't tell the difference between schemeless that we know is going to be attached to a root at some point and schemeless that is a local file relative path.

Reply:
I really think that instantiating a ResourcePath should never involve doing any I/O, both for performance and because many use cases involve building up paths that will exist in the future but don't exist yet.

What exactly gets dicey?  Most of the path manipulation code I've seen is able to work without accessing the filesystem, by making assumptions about what is and is not a directory. e.g. if you are trying to join "a/b/c" to "d/e/f", you just assume that 'c' is a directory and return "a/b/c/d/e/f" because why else would you be trying to access files inside it?  Then in the unlikely event that someone created a file named 'c' where there was supposed to be a directory, you get an exception when you try to access it as a directory.

The really tricky cases are things like abspath() that have to know what is and isn't a symlink, but we're already deferring to the python library for those from what I've seen.

Reply:
I believe we have resolved this during standup: try not to be clever given we don't really know if the schemeless URI has any relationship to the local file system at all.